cirq-core/cirq/sim/classical_simulator.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict
+from collections import defaultdict
+from cirq.sim.simulator import SimulatesSamples
+from cirq import ops, protocols
+from cirq.study.resolver import ParamResolver
+from cirq.circuits.circuit import AbstractCircuit
+from cirq.ops.raw_types import Qid
+import numpy as np
+
+
+class ClassicalStateSimulator(SimulatesSamples):
+    """basic simulator that only accepts cirq.X, cirq.ISwap, and cirq.CNOT gates
+
+    Run a simulation, mimicking quantum hardware.
+
+      Args:
+          circuit: The circuit to simulate.
+          param_resolver: Parameters to run with the program.
+          repetitions: Number of times to repeat the run. It is expected that
+              this is validated greater than zero before calling this method.
+
+      Returns:
+          A dictionary from measurement gate key to measurement
+          results. Measurement results are stored in a 3-dimensional
+          numpy array, the first dimension corresponding to the repetition.
+          the second to the instance of that key in the circuit, and the
+          third to the actual boolean measurement results (ordered by the
+          qubits being measured.)
+
+     Raises:
+      ValuesError: gate is not a cirq.XGate or cirq.Cnot
+    """
+
+    def _run(
+        self, circuit: AbstractCircuit, param_resolver: ParamResolver, repetitions: int
+    ) -> Dict[str, np.ndarray]:
+        results_dict: Dict[str, np.ndarray] = {}
+        values_dict: Dict[Qid, int] = defaultdict(int)
+        param_resolver = param_resolver or ParamResolver({})
+        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+
+        for moment in resolved_circuit:
+            for op in moment:
+                gate = op.gate
+                if isinstance(gate, ops.XPowGate) and gate.exponent == 1:
+                    values_dict[op.qubits[0]] = 1 - values_dict[op.qubits[0]]
+
+                elif isinstance(gate, ops.CNotPowGate) and gate.exponent == 1:
+                    if values_dict[op.qubits[0]] == 1:
+                        values_dict[op.qubits[1]] = 1 - values_dict[op.qubits[1]]
```

Comment:
We also need to check for `globa_shift == 0` for eigen gates -- let's instead do an equality comparision to `ops.X`, `ops.CX` etc. since eventually we'd anyway want to shift to a protocols based approach.